sim/core/item_swaps.go

```diff
@@ -0,0 +1,247 @@
+package core
+
+import (
+	"time"
+
+	"github.com/wowsims/wotlk/sim/core/proto"
+	"github.com/wowsims/wotlk/sim/core/stats"
+)
+
+type OnSwapItem func(*Simulation)
+
+const offset = proto.ItemSlot_ItemSlotMainHand
+
+type ItemSwap struct {
+	character       *Character
+	onSwapCallbacks []OnSwapItem
+
+	mhCritMultiplier     float64
+	ohCritMultiplier     float64
+	rangedCritMultiplier float64
+
+	//Used for resetting
+	initialEquippedItems   [3]Item
+	initialUnequippedItems [3]Item
+
+	//holds items that are currently not equipped
+	unEquippedItems [3]Item
+
+	//for handling 2Handers, if the unEquippedItems holds a 2Hander
+	has2H bool
+}
+
+func (character *Character) RegisterOnItemSwap(callback OnSwapItem) {
+	if character == nil || character.ItemSwap.IsEnabled() {
+		return
+	}
+
+	character.ItemSwap.onSwapCallbacks = append(character.ItemSwap.onSwapCallbacks, callback)
+}
+
+//Helper for handling Effects that use PPMManager to toggle the effect on/off
+func (swap *ItemSwap) RegisterOnSwapItemForEffectWithPPMManager(effectID int32, ppm float64, ppmm *PPMManager, aura *Aura) {
+	if swap.IsEnabled() {
+		return
+	}
+
+	character := swap.character
+	character.RegisterOnItemSwap(func(sim *Simulation) {
+		mh := character.Equip[proto.ItemSlot_ItemSlotMainHand].Enchant.EffectID == effectID
+		oh := character.Equip[proto.ItemSlot_ItemSlotOffHand].Enchant.EffectID == effectID
+
+		procMask := GetMeleeProcMaskForHands(mh, oh)
+		*ppmm = character.AutoAttacks.NewPPMManager(ppm, procMask)
+
+		if ppmm.Chance(procMask) == 0 {
+			aura.Deactivate(sim)
+		} else {
+			aura.Activate(sim)
+		}
+	})
+
+}
+
+//Helper for handling Effects that use the Aura to toggle the effect on and off
+func (swap *ItemSwap) ReigsterOnSwapItemForEffect(effectID int32, aura *Aura) {
+	if swap.IsEnabled() {
+		return
+	}
+
+	character := swap.character
+	character.RegisterOnItemSwap(func(sim *Simulation) {
+		mh := character.Equip[proto.ItemSlot_ItemSlotMainHand].Enchant.EffectID == effectID
+		oh := character.Equip[proto.ItemSlot_ItemSlotOffHand].Enchant.EffectID == effectID
+
+		if !mh && !oh {
+			aura.Deactivate(sim)
+		} else {
+			aura.Activate(sim)
+		}
+	})
+}
+
+func (character *Character) EnableItemSwap(itemSwap *proto.ItemSwap, mhCritMultiplier float64, ohCritMultiplier float64, rangedCritMultiplier float64) {
```

Comment:
All the extra parameters here and the code in multiple places for handling the Weapon struct is really messy. It's OK for now if you want to get this submitted soon, but we'll need to figure out something cleaner as this will be quite error-prone.